src/server/acl/validator.cc

```diff
@@ -31,12 +35,51 @@ namespace dfly::acl {
 
 [[nodiscard]] bool IsUserAllowedToInvokeCommandGeneric(uint32_t acl_cat,
                                                        const std::vector<uint64_t>& acl_commands,
+                                                       const AclKeys& keys, CmdArgList tail_args,
                                                        const facade::CommandId& id) {
   const auto cat_credentials = id.acl_categories();
   const size_t index = id.GetFamily();
   const uint64_t command_mask = id.GetBitIndex();
   DCHECK_LT(index, acl_commands.size());
-  return (acl_cat & cat_credentials) != 0 || (acl_commands[index] & command_mask) != 0;
+
+  const bool is_read_command = id.opt_mask() & CO::CommandOpt::READONLY;
+  const bool is_write_command = id.opt_mask() & CO::CommandOpt::WRITE;
+  const size_t step = (id.opt_mask() & CO::CommandOpt::INTERLEAVED_KEYS) ? 2 : 1;
+  const bool var_keys = (id.opt_mask() & CO::CommandOpt::VARIADIC_KEYS);
+
+  auto match = [](const auto& pattern, const auto& target) {
+    return stringmatchlen(pattern.data(), pattern.size(), target.data(), target.size(), 0);
+  };
+
+  bool key_allowed = true;
+  if (!keys.all_keys && id.first_key_pos() != 0 && (is_read_command || is_write_command)) {
+    const size_t end = var_keys ? tail_args.size() : id.last_key_pos() + 1;
+    for (size_t i = id.first_key_pos(); i < end; i += step) {
```

Comment:
there is something called a bonus key, you'll see when you use DetermineKeys()

Reply:
I forgot about this and the bonus keys. Nice catch!